### DISCUSSION TITLE
How can i validate File Input with zod?

### DISCUSSION THREAD

**Discussion by niklasgrewe** [blitz-js/blitz#2292] [Q&A]
Hi, i want to validate a File Input like this:

```js
export const CreateTeam = z.object({
  teamProfileImage: z.instanceof(File)
})
```
when i try to use it, i get the runtime error:

```
Server Error
ReferenceError: File is not defined
```
How can i solve this? I know that the validation was used both, on client and on server. I think that this is the problem, because nodejs don't know about File Type, isn't it?

**Comment by flybayer** [comment]
Are you trying to upload a file with a mutation? Mutations don't yet support file upload: https://github.com/blitz-js/blitz/discussions/2270

**Reply by niklasgrewe** [reply]
good to know, hope that will be possible soon. At the moment I only wanted to use zod for validation. How to fix my error?

**Reply by flybayer** [reply]
is that error in the browser or on the server?

**Reply by niklasgrewe** [reply]
@flybayer i think on the server. On my terminal i get this error: `ReferenceError: File is not defined` and in the browser window i get the message: `Internal Server Error`

**Reply by flybayer** [reply]
Can you post call your code related to this? Hard to help without it

**Reply by niklasgrewe** [reply]
@flybayer 

i am using this validation type 
```js
export const UserSetup = z.object({
   avatar: z.instanceof(File),
   username: z.string().min(4).max(20)
})
```
inside my Formik Form like this:
```js
<Form
  schema={UserSetup}
  initialValues={{ avatar: undefined, username: "" }}
  submitText="Take me to workspace"
  onSubmit={async (values, actions) => {
     console.log("form data", values)
  }}
>
  <UploadField />
  <InputField />
</Form>
```
inside my `UploadField` Comp, i am saving the file like this:
``` js
const handleFileChange = async (event: ChangeEvent<HTMLInputElement>) => {
   if (!event.currentTarget.files) return
   const file = event.currentTarget.files[0]
   form.setValue(file)
}
```
this is currently all the code I use. But when i refresh the page, in the browser window i get the error: `Internal Server Error` and inside my server console (inside VS Code) i am getting the error: `ReferenceError: File is not defined`

hope this is helpful to help me to solve this kind of issue

**Reply by flybayer** [reply]
Ok probably the schema code is running on the server somehow and `File` does not exist in node. 

If you need help solving that, then make a codesandbox or a git repo with reproduction of this and I can take a look.

**Reply by niklasgrewe** [reply]
@flybayer yeah the schema code is running also on the server, when i not use the schema in mutations or queries. In my case it's quite stupid, because I need the schema only client-side. Here we are again at the point you suggested yesterday, namely that we define convention of `filename.server.js` for server only files. This could also be used to fix such errors, couldn't it? I am not one of the constributors of Blitz, but I suspect that the schemas are also included in the server bundle and this is what generates the error message.

**Reply by flybayer** [reply]
Most code in your app runs on the server during build.

The way to fix is like this:

```ts
export const UserSetup = z.object({
   avatar: typeof window === 'undefined' ? z.any() : z.instanceof(File),
   username: z.string().min(4).max(20)
})
```

**Reply by bryanprimus** [reply]
> The way to fix is like this:

if we use this we will be losing the typescript type. zod will detect the type as any

**Reply by bryanprimus** [reply]
I prefer to use `z.null()` or `z.undefined()` instead of `z.any()`

**Reply by AdsonCicilioti** [reply]
Yes!  `z.any` interferes with type inference.

**Reply by kuldeep-tatvic** [reply]
Thanks @flybayer , You saved my time
